docs/design-notes/cql-extensions-internal.md

```diff
@@ -0,0 +1,39 @@
+# Scylla CQL extensions
+
+Scylla extends the CQL language to provide a few extra features. This document
+lists those extensions.
+
+## SCYLLA_TIMEUUID_LIST_INDEX function
+
+The `SCYLLA_TIMEUUID_LIST_INDEX` function allows using the raw (sstable) collection
+key for lists, which is a timeuuid, as "index" when performing a modification.
+
+The function is used to wrap the index part of a list assignment:
+
+    UPDATE ... SET list_column[SCYLLA_TIMEUUID_LIST_INDEX(:v1)] = :v2 ... 
+
+It is only used by sstableloader.
+
+## SCYLLA_COUNTER_SHARD_LIST function
+
+The `SCYLLA_COUNTER_SHARD_LIST` function allows setting a counter column value by 
+its internal representation.
+
+The function is used to wrap the value part of a counter assignment:
+
+    UPDATE ... SET counter_column = SCYLLA_COUNTER_SHARD_LIST(:v1) ... 
+
+It is only used by sstableloader.
+
+## SCYLLA_CLUSTERING_BOUND function
+
+The `SCYLLA_CLUSTERING_BOUND` function marks a query bound as being 
+a CQL clustering boundary, not a value range bound as in normal CQL.
+
+This function can be used on the right-hand side of a `WHERE` statement multicolumn 
+comparison, excluding the IN and LIKE operators. Its parameters are values corresponding 
+to the clustering prefix specified on the left-hand side of that comparison.
```

Comment:
Just thought of one more thing: we should prohibit mixing multi-column comparisons of different types (eg, `(c1,c2) > SCYLLA_CLUSTERING_BOUND(1,2) AND (c1,c2)<(10,20)`).

Reply:
Why? While it makes no sense, the parser should handle it just fine. The end result will be junk, but that is what caller asked for.

Reply:
I don't think the end result would be junk; it's much more likely that the first restriction's mode will be obeyed and the second ignored.

The reason I would prohibit such mixing is to make it easier to debug obscure errors when the mixup happens unintentionally.  I can't imagine it would ever happen deliberately, so it's better to reject it and help the user realize their error.

Reply:
Afaict, it would generate two sets of restrictions that are eventually merged in `do_merge_with`. The result should not be ignored, it would just have somewhat hard to think of bounds. Example (from test - updating it):
```
-- mixing apples and make_count_rows_function
SELECT * FROM foo WHERE a=0 AND (b, c) > SCYLLA_CLUSTERING_BOUND(2, 0)  AND (b, c) < (1, 1);
{
        "rows" : 
        [
                {
                        "a" : "0",
                        "b" : "1",
                        "c" : "0"
                }
        ]
}
-- and again --
SELECT * FROM foo WHERE a=0 AND (b, c) > SCYLLA_CLUSTERING_BOUND(2, 0)  AND (b, c) < (2, 0);
{
        "rows" : null
}
```

Reply:
> Afaict, it would generate two sets of restrictions that are eventually merged in `do_merge_with`.

... which [produces](https://sourcegraph.com/github.com/scylladb/scylla/-/blob/cql3/restrictions/multi_column_restriction.hh#L436:18) a single `multi_column_restriction::slice` object, with a single `mode`.

> SELECT * FROM foo WHERE a=0 AND (b, c) > SCYLLA_CLUSTERING_BOUND(2, 0)  AND (b, c) < (1, 1);
> {
>         "rows" : 
>         [
>                 {
>                         "a" : "0",
>                         "b" : "1",
>                         "c" : "0"
>                 }

Don't you think this should include (0,0,0) if both modes were correctly obeyed?

Reply:
Oh, I did not say they would be correctly obeyed. Just that it will "work", as in produce garbage result. But this is an internal feature. I don't think it needs to be idiot-proof and self-validating. Esp. since checking in grammar would be terrible, and checking in the restriction might be error-prone.

Reply:
Actually, re-reading it, the check can probably just be that two merged slices must have same "mode", otherwise -> exception. So maybe I'm changing my mind towards adding a check-fail.

Reply:
Yes, even for an internal feature it's really undesirable that "A and B" produces a different result from "B and A".  The check can be done in `multi_column_restriction::slice::do_merge_with` linked above -- just compare the `mode` members of the two objects and raise a flag if they differ.

Reply:
Updated with a check + exception (+ test cases)